### ISSUE TITLE
Extracting different features

### ISSUE THREAD

**Issue by casbolwerk** [david8862/keras-YOLOv3-model-set#134]
While playing around with the YOLOv4 mobilenet model, I was trying to extract different features. 

By default, the code extracts layers `conv_pw_13_relu`, `conv_pw_11_relu` and `conv_pw_5_relu` from line 23 onward in `yolo4_mobilenet.py`. To try something different, I wanted to extract a different layer. The problem I think starts with the difference in channel numbers, which causes some incompatibility. When extracting `conv_pw_9_relu` instead of `conv_pw_5_relu`, the number of channel numbers must change from 256 to 512 according to the model summary of mobilenet.

```
...
_________________________________________________________________
conv_pw_5 (Conv2D)           (None, None, None, 256)   65536
_________________________________________________________________
conv_pw_5_bn (BatchNormaliza (None, None, None, 256)   1024
_________________________________________________________________
conv_pw_5_relu (ReLU)        (None, None, None, 256)   0
_________________________________________________________________

...
_________________________________________________________________
conv_pw_9 (Conv2D)           (None, None, None, 512)   262144
_________________________________________________________________
conv_pw_9_bn (BatchNormaliza (None, None, None, 512)   2048
_________________________________________________________________
conv_pw_9_relu (ReLU)        (None, None, None, 512)   0
_________________________________________________________________
...
```
So I tried extracting a different layer like this (change made in `yolo4_mobilenet.py`).
```
    # f1: 13 x 13 x (1024*alpha) for 416 input
    f1 = mobilenet.get_layer('conv_pw_13_relu').output
    # f2: 26 x 26 x (512*alpha) for 416 input
    f2 = mobilenet.get_layer('conv_pw_11_relu').output
    # f3: 52 x 52 x (256*alpha) for 416 input
    #f3 = mobilenet.get_layer('conv_pw_5_relu').output
    f3 = mobilenet.get_layer('conv_pw_9_relu').output
    #print(f3.shape)

    f1_channel_num = int(1024*alpha)
    f2_channel_num = int(512*alpha)
    f3_channel_num = int(512*alpha)
    #f3_channel_num = int(256*alpha)
```
But this throws an error once the model starts training.
```
File "<string>", line 3, in raise_from
tensorflow.python.framework.errors_impl.InvalidArgumentError:  ConcatOp : Dimensions of inputs should match: shape[0] = [8,26,26,256] vs. shape[1] = [8,52,52,256]
         [[node model_2/concatenate_7/concat (defined at train.py:173) ]] [Op:__inference_distributed_function_33236]

Function call stack:
distributed_function
```
Line 173 of train.py is the model.fit_generator call. It says something about the concatenate_7 layer, which basically concatenates the `conv_pw_9_relu` layer with previous concatenations. This seems to be incompatible shape-wise. The yolo model summary shows this.
```
...
__________________________________________________________________________________________________
conv2d_48 (Conv2D)              (None, None, None, 2 65536       leaky_re_lu_47[0][0]
__________________________________________________________________________________________________
conv2d_49 (Conv2D)              (None, None, None, 2 131072      conv_pw_9_relu[0][0]
__________________________________________________________________________________________________
batch_normalization_48 (BatchNo (None, None, None, 2 1024        conv2d_48[0][0]
__________________________________________________________________________________________________
batch_normalization_49 (BatchNo (None, None, None, 2 1024        conv2d_49[0][0]
__________________________________________________________________________________________________
leaky_re_lu_48 (LeakyReLU)      (None, None, None, 2 0           batch_normalization_48[0][0]
__________________________________________________________________________________________________
leaky_re_lu_49 (LeakyReLU)      (None, None, None, 2 0           batch_normalization_49[0][0]
__________________________________________________________________________________________________
up_sampling2d_3 (UpSampling2D)  (None, None, None, 2 0           leaky_re_lu_48[0][0]
__________________________________________________________________________________________________
concatenate_7 (Concatenate)     (None, None, None, 5 0           leaky_re_lu_49[0][0]
                                                                 up_sampling2d_3[0][0]
__________________________________________________________________________________________________
...
```

In short, I think it should be possible to extract features from different layers but don't understand how. Mobilenetv2 extracts different features for example, that don't have this property of halving the channel numbers each time.
The problem doesn't seem to be in `yolo4_predictions` (layers.py) or `yolo4_mobilenet_body` (yolo4_mobilenet.py)

@david8862 can you show how to do this?

**Comment by casbolwerk** [comment]
One thing I found out, is that the incompatibility is more due to this.
```
    # input: 416 x 416 x 3
    # conv_pw_13_relu :13 x 13 x (1024*alpha)
    # conv_pw_11_relu :26 x 26 x (512*alpha)
    # conv_pw_5_relu : 52 x 52 x (256*alpha)
```
These 13, 26 and 52 values are the cause of the incompatibility. If I extract features from the 9th conv pw layer it works for example.
```
    # f1: 13 x 13 x (1024*alpha) for 416 input
    f1 = mobilenet.get_layer('conv_pw_13_relu').output
    # f2: 26 x 26 x (512*alpha) for 416 input
    #f2 = mobilenet.get_layer('conv_pw_11_relu').output
    f2 = mobilenet.get_layer('conv_pw_9_relu').output
    # f3: 52 x 52 x (256*alpha) for 416 input
    f3 = mobilenet.get_layer('conv_pw_5_relu').output
    #print(f3.shape)

    f1_channel_num = int(1024*alpha)
    f2_channel_num = int(512*alpha)
    f3_channel_num = int(256*alpha)
```
Now my question becomes, is it necessary for these values to be this way? Or is it possible to change them in some way, so that I can still extract as in the example from the issue above?

**Comment by david8862** [comment]
1. In your 1st failed case, the feature size of 'conv_pw_9_relu' has been 26x26 instead of 52x52 for 'conv_pw_5_relu', which doesn't match the FPN "Upsample -> Concatenate" structure. That's the root cause of error.
2. The "(None None)" feature shape show in summary is caused by dynamic input shape at input layer. You can try to change [here](https://github.com/david8862/keras-YOLOv3-model-set/blob/master/yolo3/model.py#L179) to use a static input shape (e.g. 416x416) to figure out the middle layer's shape, and then which layer could be picked for which FPN head.

**Comment by casbolwerk** [comment]
Thanks for the answer, the static input shape really helps to clear things up.

It's easy now to change out layers that have the same feature size. But I still wanted to try to get it working for layers with different feature sizes, so I decided to modify yolo4_predictions. What I basically did was removing up- and downsampling that I didn't need because of identical feature sizes. At least I think it's just for getting layers to have the same feature size shape.

So as an example I changed out
```
x2_upsample = compose(
             DarknetConv2D_BN_Leaky(f3_channel_num//2, (1,1)),
             UpSampling2D(2))(x2)
...
x3 = Concatenate()([x3, x2_upsample])
```
for
```
x2 = DarknetConv2D_BN_Leaky(f3_channel_num//2, (1,1))(x2)
...
x3 = Concatenate()([x3, x2])
```
This seems to work fine.

I'm more unsure about downsampling, where I did this
```
x3_downsample = compose(
            ZeroPadding2D(((1,0),(1,0))),
            DarknetConv2D_BN_Leaky(f2_channel_num//2, (3,3), strides=(2,2)))(x3)

x2 = Concatenate()([x3, x2])
```
for
```
x3 = DarknetConv2D_BN_Leaky(f2_channel_num//2, (1,1))(x3)

x2 = Concatenate()([x3, x2])
```
So here I don't take the strides into account and change (3,3) to (1,1). I'm not sure at all if this is correct, but it didn't give immediate errors. When I left the conv layer the same as it was, it still changed the feature size.

The predictions part of the model seems fine now, but I think the loss part of the model expects the same feature sizes still. It gives this error.
```
tensorflow.python.framework.errors_impl.InvalidArgumentError:  Incompatible shapes: [8,52,52,3,1] vs. [8,26,26,3,1]
         [[node model_2/yolo_loss/logistic_loss_8/mul (defined at D:\Program Files\keras-YOLOv3-model-set\yolo3\loss.py:336) ]] [Op:__inference_distributed_function_33083]

Errors may have originated from an input operation.
Input Source operations connected to node model_2/yolo_loss/logistic_loss_8/mul:
 model_2/yolo_loss/strided_slice_76 (defined at D:\Program Files\keras-YOLOv3-model-set\yolo3\loss.py:287)
```
So I think the yolo layer expects the feature sizes to be the default ones, but I'm not sure how to fix this. The last layers of the yolo model now look like this.
```
...
__________________________________________________________________________________________________
predict_conv_1 (Conv2D)         (None, 13, 13, 18)   18450       leaky_re_lu_69[0][0]
__________________________________________________________________________________________________
predict_conv_2 (Conv2D)         (None, 26, 26, 18)   9234        leaky_re_lu_62[0][0]
__________________________________________________________________________________________________
predict_conv_3 (Conv2D)         (None, 26, 26, 18)   9234        leaky_re_lu_55[0][0]
__________________________________________________________________________________________________
y_true_0 (InputLayer)           [(None, None, None,  0
__________________________________________________________________________________________________
y_true_1 (InputLayer)           [(None, None, None,  0
__________________________________________________________________________________________________
y_true_2 (InputLayer)           [(None, None, None,  0
__________________________________________________________________________________________________
yolo_loss (Lambda)              ((1,), (), (), ())   0           predict_conv_1[0][0]
                                                                 predict_conv_2[0][0]
                                                                 predict_conv_3[0][0]
                                                                 y_true_0[0][0]
                                                                 y_true_1[0][0]
                                                                 y_true_2[0][0]

__________________________________________________________________________________________________
...
```
As you can see, the conv predict layers have the feature sizes I want them to have. And the yolo_loss layer doesn't like it. Is it a problem with y_true?

**Comment by david8862** [comment]
@casbolwerk  Yes. For changing the default (input/32, input/16, input/8) feature size, you need to also adjust & fitting the y_true shape in [preprocess_true_boxes](https://github.com/david8862/keras-YOLOv3-model-set/blob/master/yolo3/data.py#L87) and corresponding grid shapes in [yolo3_loss](https://github.com/david8862/keras-YOLOv3-model-set/blob/master/yolo3/loss.py#L248), and may also some related shapes around. New anchors are also needed for the feature size.

**Comment by casbolwerk** [comment]
Ah okay, that's a bit over my head. Thanks though!

**Comment by casbolwerk** [comment]
Hi @david8862 

Sorry for reopening, but I think I misunderstood you. It appears to be the case that the things that you mention in the quoted comment need to be made for all extracted layers. I thought it only had to be changed for extracted layers where the feature sizes change. This doesn't seem to be the case.

> For changing the default (input/32, input/16, input/8) feature size, you need to also adjust & fitting the y_true shape in preprocess_true_boxes and corresponding grid shapes in yolo3_loss, and may also some related shapes around. New anchors are also needed for the feature size.

Just now I trained a new model, where instead of extracting layers 13, 11 and 5 I extracted layers 12, 8 and 5. The predictions are horrible. Though the loss is similar to previous models, the mAP is not even close.
Because of this, I reached the conclusion that changes should be made to default parameters so that they match these new layers.

So should grid shapes (apparently along with y_true shape) and anchors be changed for this? If yes, could you give some explanation for doing this?

**Comment by david8862** [comment]
You mean using "conv_pw_12_relu", "conv_pw_8_relu", "conv_pw_5_relu"? From the feature map shape it should be compatible with current model structure, so no need to change grid & anchors

**Comment by casbolwerk** [comment]
Okay, that's what I though. Then it might just be bad performance because of the change in layers, not configuration. Thanks!
